src/TolerantDefinitionResolver.php

```diff
@@ -0,0 +1,880 @@
+        if ($node instanceof Node\Expr\Variable) {
+            // Resolve $this
+            if ($node->name === 'this' && $fqn = $this->getContainingClassFqn($node)) {
+                return $this->index->getDefinition($fqn, false);
+            }
+            // Resolve the variable to a definition node (assignment, param or closure use)
+            $defNode = self::resolveVariableToNode($node);
+            if ($defNode === null) {
+                return null;
+            }
+            return $this->createDefinitionFromNode($defNode);
+        }
+        // Other references are references to a global symbol that have an FQN
+        // Find out the FQN
+        $fqn = $this->resolveReferenceNodeToFqn($node);
+        if ($fqn === null) {
+            return null;
+        }
+        // If the node is a function or constant, it could be namespaced, but PHP falls back to global
+        // http://php.net/manual/en/language.namespaces.fallback.php
+        $parent = $node->getAttribute('parentNode');
+        $globalFallback = $parent instanceof Node\Expr\ConstFetch || $parent instanceof Node\Expr\FuncCall;
+        // Return the Definition object from the index index
+        return $this->index->getDefinition($fqn, $globalFallback);
+    }
+
+    /**
+     * Returns all possible FQNs in a type
+     *
+     * @param Type $type
+     * @return string[]
+     */
+    public static function getFqnsFromType(Type $type): array
+    {
+        $fqns = [];
+        if ($type instanceof Types\Object_) {
+            $fqsen = $type->getFqsen();
+            if ($fqsen !== null) {
+                $fqns[] = substr((string)$fqsen, 1);
+            }
+        }
+        if ($type instanceof Types\Compound) {
+            for ($i = 0; $t = $type->get($i); $i++) {
+                foreach (self::getFqnsFromType($type) as $fqn) {
+                    $fqns[] = $fqn;
+                }
+            }
+        }
+        return $fqns;
+    }
+
+    /**
+     * Given any node, returns the FQN of the symbol that is referenced
+     * Returns null if the FQN could not be resolved or the reference node references a variable
+     *
+     * @param Node $node
+     * @return string|null
+     */
+    public function resolveReferenceNodeToFqn(Node $node)
+    {
+        $parent = $node->getAttribute('parentNode');
+
+        if (
+            $node instanceof Node\Name && (
+                $parent instanceof Node\Stmt\ClassLike
+                || $parent instanceof Node\Param
+                || $parent instanceof Node\FunctionLike
+                || $parent instanceof Node\Stmt\GroupUse
+                || $parent instanceof Node\Expr\New_
+                || $parent instanceof Node\Expr\StaticCall
+                || $parent instanceof Node\Expr\ClassConstFetch
+                || $parent instanceof Node\Expr\StaticPropertyFetch
+                || $parent instanceof Node\Expr\Instanceof_
```

Comment:
That might be an interesting optimization to make later, if I can collect data on which structures are used most often.